documentation/proposals/Proposal - Axis Input Devices.md

```diff
@@ -0,0 +1,611 @@
+This proposal can be thought of in two primary areas: the API surface, and the implementation requirements. Wherever it was possible, the API surface was designed to enforce certain requirements, however there are quite a few requirements that must be followed that are not enforced by the API itself. Comments have been added to the API where these requirements are necessary and further detail is provided where needed. Feedback on clarity is welcome.
+
+# Modifications to the Multi-Backend Input Proposal
+We propose the following properties be added to `IInputDevice` for their general utility and to help keep this API and the [base input API](./Proposal - Multi-Backend
+Input.md) consistent with each other.
+
+```csharp
+public interface IInputDevice : IEquatable<IInputDevice>
+{
+    // ---- Current API ----- //
+    nint Id { get; } 
+    string Name { get; }
+    // ---------------------- //
+
+    /// <summary>
+    /// A human-readable description of the device, intended for end-user display
+    /// </summary>
+    public string? Description => null;
+
+    /// <summary>
+    /// A human-readable description of the device's status, intended for end-user display
+    /// In the case of device malfunctions, especially those that are composed of multiple devices,
+    /// this can be used to display any issues preventing the device from functioning properly
+    /// </summary>
+    public string? Status => null;
+    
+    /// <summary>
+    /// False if this is a single real-world device, true if this is a device based on one or more physical devices,
+    /// or is entirely virtual (e.g. a pointer in a VR environment, a virtual gamepad, etc)
+    /// </summary>
+    bool IsVirtual => false;
+}
+```
+
+# Core API
+
+The root of the API is in the interface `IAxisDevice`. This interface is essentially a host for collections of axes paired with definitions of these axes and how they relate to one another. 
+
+It is important to note that the value of an input axis a single 32-bit floating point value. is primarily expected between 0-1.0. There are some axis types, or rather `AxisTrait`s, that are not straightforwardly normalized, and those respective axes can be labeled as such (more on that later).
+
+## Device Validation
+Due to the complex set of constraints detailed below, it is important that a validation layer exists to ensure that higher-level code can trust the device is following important conventions and constraints. As a result, each `IAxisDevice` must be validated, similar to Vulkan's validation layers, through the current `IInputContext` if such a context desires to make reliable use of this API. The Silk.NET team shall make their validation implementation easily callable by custom `IInputContext` implementations due to its inevitable complexity.
+
+The signature of Silk's validation method should resemble the following:
+
+```csharp
+public static bool IsAxisDeviceValid(IAxisDevice device, [NotNullWhen(false)] out string? error);

```

Comment:
```suggestion
namespace Silk.NET.Input;

public partial interface IAxisDevice
{
    static sealed void Validate(IAxisDevice device);
    static sealed bool TryValidate(IAxisDevice device);
}
```
We can bikeshed these methods if we want, I just don't like the opaque error string as you can't really programmatically rely on its value given it could be anything. It's an exceptional circumstance that a device is invalid, hence this should be given in exceptions, unless for some reason the user is tolerant of invalid device in which case we expose the try-style method (but no further info as again this could be anything...)

Reply:
I'm personally happy with an exception - just didnt know just how core you were willing to take this (e.g. it MUST be correct or it SHOULD be correct but some implementations might lag behind etc etc)